File: Shared/RepoPressCoreContracts/dotnet/src/RepoPress.Core/FrontMatterDocumentRenderer.cs
using System;
using System.Collections.Generic;

namespace RepoPress.Core
{
    public enum FrontMatterDocumentSyntax
    {
        Yaml,
        Toml
    }

    public enum FrontMatterTaxonomyLayout
    {
        InlineTable,
        Table
    }

    public sealed class FrontMatterDocument
    {
        public FrontMatterDocument(
            FrontMatterDocumentSyntax syntax,
            string title,
            string formattedDate,
            string? slug,
            bool? draftFlag,
            string? summaryField,
            string? summary,
            IReadOnlyList<string> authors,
            IReadOnlyList<string> tags,
            IReadOnlyList<string> categories,
            FrontMatterTaxonomyLayout taxonomyLayout,
            string? coverField,
            string? coverPath,
            bool writesCoverInExtraTable,
            string bodyMarkdown)
        {
            Syntax = syntax;
            Title = title ?? throw new ArgumentNullException(nameof(title));
            FormattedDate = formattedDate ?? throw new ArgumentNullException(nameof(formattedDate));
            Slug = slug;
            DraftFlag = draftFlag;
            SummaryField = summaryField;
            Summary = summary;
            Authors = CopyCollection(authors, nameof(authors));
            Tags = CopyCollection(tags, nameof(tags));
            Categories = CopyCollection(categories, nameof(categories));
            TaxonomyLayout = taxonomyLayout;
            CoverField = coverField;
            CoverPath = coverPath;
            WritesCoverInExtraTable = writesCoverInExtraTable;
            BodyMarkdown = bodyMarkdown ?? throw new ArgumentNullException(nameof(bodyMarkdown));
        }

        public FrontMatterDocumentSyntax Syntax { get; }

        public string Title { get; }

        public string FormattedDate { get; }

        public string? Slug { get; }

        public bool? DraftFlag { get; }

        public string? SummaryField { get; }

        public string? Summary { get; }

        public IReadOnlyList<string> Authors { get; }

        public IReadOnlyList<string> Tags { get; }

        public IReadOnlyList<string> Categories { get; }

        public FrontMatterTaxonomyLayout TaxonomyLayout { get; }

        public string? CoverField { get; }

        public string? CoverPath { get; }

        public bool WritesCoverInExtraTable { get; }

        public string BodyMarkdown { get; }

        private static IReadOnlyList<string> CopyCollection(
            IReadOnlyList<string> values,
            string parameterName)
        {
            if (values == null)
            {
                throw new ArgumentNullException(parameterName);
            }

            var copy = new List<string>(values.Count);
            for (int index = 0; index < values.Count; index++)
            {
                copy.Add(values[index] ?? throw new ArgumentNullException(parameterName));
            }

            return copy.AsReadOnly();
        }
    }

    public sealed class FrontMatterDocumentRenderer
    {
        public string Render(FrontMatterDocument document)
        {
            if (document == null)
            {
                throw new ArgumentNullException(nameof(document));
            }

            return document.Syntax == FrontMatterDocumentSyntax.Yaml
                ? RenderYaml(document)
                : RenderToml(document);
        }

        public string MarkdownDocument(FrontMatterDocument document)
        {
            if (document == null)
            {
                throw new ArgumentNullException(nameof(document));
            }

            return Render(document) + "\n\n" + document.BodyMarkdown.Trim() + "\n";
        }

        private static string RenderYaml(FrontMatterDocument document)
        {
            var lines = new List<string>
            {
                "---",
                "title: " + Quoted(document.Title),
                "date: " + Quoted(document.FormattedDate)
            };
            AppendCommonScalarFields(lines, document, ":");
            AppendYamlCollections(lines, document);
            if (document.DraftFlag.HasValue)
            {
                lines.Add("draft: " + (document.DraftFlag.Value ? "true" : "false"));
            }

            AppendSummary(lines, document, ":");
            AppendCover(lines, document, ":");
            lines.Add("---");
            return string.Join("\n", lines);
        }

        private static string RenderToml(FrontMatterDocument document)
        {
            var lines = new List<string>
            {
                "+++",
                "title = " + Quoted(document.Title),
                "date = " + document.FormattedDate
            };
            AppendCommonScalarFields(lines, document, "=");
            if (document.DraftFlag.HasValue)
            {
                lines.Add("draft = " + (document.DraftFlag.Value ? "true" : "false"));
            }

            AppendSummary(lines, document, "=");
            if (document.Authors.Count > 0)
            {
                lines.Add("authors = [" + QuotedList(document.Authors) + "]");
            }

            if (!document.WritesCoverInExtraTable)
            {
                AppendCover(lines, document, "=");
            }

            AppendTomlTaxonomies(lines, document);
            if (document.WritesCoverInExtraTable && document.CoverPath != null)
            {
                lines.Add("[extra]");
                lines.Add("og_preview_img = " + Quoted(document.CoverPath));
            }

            lines.Add("+++");
            return string.Join("\n", lines);
        }

        private static void AppendCommonScalarFields(
            List<string> lines,
            FrontMatterDocument document,
            string separator)
        {
            if (document.Slug != null)
            {
                lines.Add(Assignment("slug", Quoted(document.Slug), separator));
            }
        }

        private static void AppendYamlCollections(
            List<string> lines,
            FrontMatterDocument document)
        {
            if (document.Tags.Count > 0)
            {
                lines.Add("tags: [" + QuotedList(document.Tags) + "]");
            }

            if (document.Categories.Count > 0)
            {
                lines.Add("categories: [" + QuotedList(document.Categories) + "]");
            }

            if (document.Authors.Count > 0)
            {
                lines.Add("authors: [" + QuotedList(document.Authors) + "]");
            }
        }

        private static void AppendSummary(
            List<string> lines,
            FrontMatterDocument document,
            string separator)
        {
            if (document.SummaryField != null && document.Summary != null)
            {
                lines.Add(Assignment(document.SummaryField, Quoted(document.Summary), separator));
            }
        }

        private static void AppendCover(
            List<string> lines,
            FrontMatterDocument document,
            string separator)
        {
            if (document.CoverField != null && document.CoverPath != null)
            {
                lines.Add(Assignment(document.CoverField, Quoted(document.CoverPath), separator));
            }
        }

        private static void AppendTomlTaxonomies(
            List<string> lines,
            FrontMatterDocument document)
        {
            if (document.Tags.Count == 0 && document.Categories.Count == 0)
            {
                return;
            }

            if (document.TaxonomyLayout == FrontMatterTaxonomyLayout.InlineTable)
            {
                var entries = new List<string>();
                if (document.Tags.Count > 0)
                {
                    entries.Add("tags = [" + QuotedList(document.Tags) + "]");
                }

                if (document.Categories.Count > 0)
                {
                    entries.Add("categories = [" + QuotedList(document.Categories) + "]");
                }

                lines.Add("taxonomies = { " + string.Join(", ", entries) + " }");
            }
            else
            {
                lines.Add("[taxonomies]");
                if (document.Tags.Count > 0)
                {
                    lines.Add("tags = [" + QuotedList(document.Tags) + "]");
                }

                if (document.Categories.Count > 0)
                {
                    lines.Add("categories = [" + QuotedList(document.Categories) + "]");
                }
            }
        }

        private static string QuotedList(IReadOnlyList<string> values)
        {
            var quoted = new string[values.Count];
            for (int index = 0; index < values.Count; index++)
            {
                quoted[index] = Quoted(values[index]);
            }

            return string.Join(", ", quoted);
        }

        private static string Assignment(string field, string value, string separator)
        {
            return separator == ":"
                ? field + ": " + value
                : field + " = " + value;
        }

        private static string Quoted(string value)
        {
            return "\""
                + value
                    .Replace("\\", "\\\\")
                    .Replace("\"", "\\\"")
                    .Replace("\n", "\\n")
                    .Replace("\r", "\\r")
                    .Replace("\t", "\\t")
                + "\"";
        }
    }
}

File: Shared/RepoPressCoreContracts/swift/Sources/RepoPressCore/FrontMatterDocumentRenderer.swift
import Foundation

public enum FrontMatterDocumentSyntax: String, Codable, Hashable, Sendable {
  case yaml
  case toml
}

public enum FrontMatterTaxonomyLayout: String, Codable, Hashable, Sendable {
  case inlineTable
  case table
}

public struct FrontMatterDocument: Codable, Hashable, Sendable {
  public var syntax: FrontMatterDocumentSyntax
  public var title: String
  public var formattedDate: String
  public var slug: String?
  public var draftFlag: Bool?
  public var summaryField: String?
  public var summary: String?
  public var authors: [String]
  public var tags: [String]
  public var categories: [String]
  public var taxonomyLayout: FrontMatterTaxonomyLayout
  public var coverField: String?
  public var coverPath: String?
  public var writesCoverInExtraTable: Bool
  public var bodyMarkdown: String

  public init(
    syntax: FrontMatterDocumentSyntax,
    title: String,
    formattedDate: String,
    slug: String?,
    draftFlag: Bool?,
    summaryField: String?,
    summary: String?,
    authors: [String],
    tags: [String],
    categories: [String],
    taxonomyLayout: FrontMatterTaxonomyLayout,
    coverField: String?,
    coverPath: String?,
    writesCoverInExtraTable: Bool,
    bodyMarkdown: String
  ) {
    self.syntax = syntax
    self.title = title
    self.formattedDate = formattedDate
    self.slug = slug
    self.draftFlag = draftFlag
    self.summaryField = summaryField
    self.summary = summary
    self.authors = authors
    self.tags = tags
    self.categories = categories
    self.taxonomyLayout = taxonomyLayout
    self.coverField = coverField
    self.coverPath = coverPath
    self.writesCoverInExtraTable = writesCoverInExtraTable
    self.bodyMarkdown = bodyMarkdown
  }
}

public struct FrontMatterDocumentRenderer: Sendable {
  public init() {}

  public func render(_ document: FrontMatterDocument) -> String {
    switch document.syntax {
    case .yaml:
      return renderYAML(document)
    case .toml:
      return renderTOML(document)
    }
  }

  public func markdownDocument(_ document: FrontMatterDocument) -> String {
    "\(render(document))\n\n\(document.bodyMarkdown.trimmingCharacters(in: .whitespacesAndNewlines))\n"
  }

  private func renderYAML(_ document: FrontMatterDocument) -> String {
    var lines = [
      "---",
      "title: \(quoted(document.title))",
      "date: \(quoted(document.formattedDate))",
    ]
    appendCommonScalarFields(to: &lines, document: document, separator: ":")
    appendYAMLCollections(to: &lines, document: document)
    if let draftFlag = document.draftFlag {
      lines.append("draft: \(draftFlag ? "true" : "false")")
    }
    appendSummary(to: &lines, document: document, separator: ":")
    appendCover(to: &lines, document: document, separator: ":")
    lines.append("---")
    return lines.joined(separator: "\n")
  }

  private func renderTOML(_ document: FrontMatterDocument) -> String {
    var lines = [
      "+++",
      "title = \(quoted(document.title))",
      "date = \(document.formattedDate)",
    ]
    appendCommonScalarFields(to: &lines, document: document, separator: "=")
    if let draftFlag = document.draftFlag {
      lines.append("draft = \(draftFlag ? "true" : "false")")
    }
    appendSummary(to: &lines, document: document, separator: "=")
    if !document.authors.isEmpty {
      lines.append("authors = [\(quotedList(document.authors))]")
    }
    if !document.writesCoverInExtraTable {
      appendCover(to: &lines, document: document, separator: "=")
    }
    appendTOMLTaxonomies(to: &lines, document: document)
    if document.writesCoverInExtraTable, let coverPath = document.coverPath {
      lines.append("[extra]")
      lines.append("og_preview_img = \(quoted(coverPath))")
    }
    lines.append("+++")
    return lines.joined(separator: "\n")
  }

  private func appendCommonScalarFields(
    to lines: inout [String],
    document: FrontMatterDocument,
    separator: String
  ) {
    if let slug = document.slug {
      lines.append(assignment(field: "slug", value: quoted(slug), separator: separator))
    }
  }

  private func appendYAMLCollections(
    to lines: inout [String],
    document: FrontMatterDocument
  ) {
    if !document.tags.isEmpty {
      lines.append("tags: [\(quotedList(document.tags))]")
    }
    if !document.categories.isEmpty {
      lines.append("categories: [\(quotedList(document.categories))]")
    }
    if !document.authors.isEmpty {
      lines.append("authors: [\(quotedList(document.authors))]")
    }
  }

  private func appendSummary(
    to lines: inout [String],
    document: FrontMatterDocument,
    separator: String
  ) {
    guard let field = document.summaryField, let summary = document.summary else {
      return
    }
    lines.append(assignment(field: field, value: quoted(summary), separator: separator))
  }

  private func appendCover(
    to lines: inout [String],
    document: FrontMatterDocument,
    separator: String
  ) {
    guard let field = document.coverField, let path = document.coverPath else {
      return
    }
    lines.append(assignment(field: field, value: quoted(path), separator: separator))
  }

  private func appendTOMLTaxonomies(
    to lines: inout [String],
    document: FrontMatterDocument
  ) {
    guard !document.tags.isEmpty || !document.categories.isEmpty else {
      return
    }
    switch document.taxonomyLayout {
    case .inlineTable:
      var entries: [String] = []
      if !document.tags.isEmpty {
        entries.append("tags = [\(quotedList(document.tags))]")
      }
      if !document.categories.isEmpty {
        entries.append("categories = [\(quotedList(document.categories))]")
      }
      lines.append("taxonomies = { \(entries.joined(separator: ", ")) }")
    case .table:
      lines.append("[taxonomies]")
      if !document.tags.isEmpty {
        lines.append("tags = [\(quotedList(document.tags))]")
      }
      if !document.categories.isEmpty {
        lines.append("categories = [\(quotedList(document.categories))]")
      }
    }
  }

  private func quotedList(_ values: [String]) -> String {
    values.map(quoted).joined(separator: ", ")
  }

  private func assignment(field: String, value: String, separator: String) -> String {
    separator == ":" ? "\(field): \(value)" : "\(field) = \(value)"
  }

  private func quoted(_ value: String) -> String {
    let escaped = value
      .replacingOccurrences(of: "\\", with: "\\\\")
      .replacingOccurrences(of: "\"", with: "\\\"")
      .replacingOccurrences(of: "\n", with: "\\n")
      .replacingOccurrences(of: "\r", with: "\\r")
      .replacingOccurrences(of: "\t", with: "\\t")
    return "\"\(escaped)\""
  }
}
